core/src/main/java/org/apache/calcite/sql/validate/implicit/TypeCoercionImpl.java

```diff
@@ -292,16 +292,17 @@ protected boolean binaryArithmeticWithStrings(
     }
 
     RelDataType commonType;
-    if (SqlTypeUtil.sameNamedType(type1, type2)) {
+    // First check when one of the types is a string (see JavaDoc on cmmonTypeForBinaryComparison)
+    commonType = commonTypeForBinaryComparison(type1, type2);
+    if (commonType == null && SqlTypeUtil.isComparable(type1, type2)) {
```

Comment:
Why do we need to judge this `SqlTypeUtil.isComparable(type1, type2)`?

Reply:
This is about comparison operators, you'd think that if 2 values with types are comparable, then the values can be compared by being cast to a common type.